RFC: Reference Target for Cross-root ARIA

behowell opened the pull request:
> [!NOTE]  
> This PR has been merged. The latest version of the explainer is available here:
> [📜 Reference Target for Cross-root ARIA](https://github.com/WICG/aom/blob/gh-pages/reference-target-explainer.md)
> 
> For further feedback, please log an issue or PR. 

Reference Target is a new feature that enables creating ARIA links to elements inside a component's shadow DOM, while maintaining encapsulation of the internal details of the shadow DOM. 

I'd really appreciate any feedback and comments! You can leave comments directly on the .md file in this PR.

Thank you to @alice, @annevk, and @chrisdholt for taking an earlier look at this proposal and giving feedback.

behowell committed:
Initial draft of reference delegate explainer

behowell committed:
Add sections, edit for clarity

behowell committed:
Add interactions with forms, and author line

behowell committed:
Minor tweaks to wording of the introduction

behowell committed:
Remove the Background section; the example didn't add much and was potentially confusing

behowell committed:
Rename section to Considered Alternatives

Westbrook commented on reference-delegate-explainer.md:
```diff
@@ -0,0 +1,502 @@
+* `aria-errormessage`
+* `aria-flowto`
+* `aria-labelledby`
+* `aria-owns`
```
Should include the `popover` and “invoker” attributes here?

behowell replied:
Good call -- yes it should include them. I've also added a few other attributes to the list that I missed earlier.

behowell committed:
Include popovertarget and several other included attributes; other edits for clarity

behowell committed:
Remove some unnecessary MDN links

behowell committed:
Add link to invokers proposal

Westbrook commented on reference-delegate-explainer.md:
````diff
@@ -0,0 +1,534 @@
+
+The equivalent declarative attribute is `shadowrootreferencedelegatemap`, which is a comma-separated list of attribute to ID mappings.
+
+```html
+<input role="combobox"
+       aria-controls="fancy-listbox"
+       aria-activedescendant="fancy-listbox" />
+<fancy-listbox id="fancy-listbox">
+  <template shadowrootmode="closed"
+            shadowrootreferencedelegatemap="aria-controls: real-listbox,
+                                            aria-activedescendant: option-2">
+    <div id="real-listbox" role="listbox">
+      <div id="option-1" role="option">Option 1</div>
+      <div id="option-2" role="option">Option 2</div>
+    </div>
+  </template>
+</fancy-listbox>
+```
````
Does this approach allow a user to do the following?
```html
<fancy-input
  role="combobox"
  aria-controls="fancy-listbox"
  aria-activedescendant="fancy-listbox"
>
  <template shaodwrootmode="open" shadowrootreferencedelegate="input">
    <input id="input" />
  </template>
</fancy-input>
<fancy-listbox id="fancy-listbox">
  <template shadowrootmode="open"
            shadowrootreferencedelegatemap="aria-controls: real-listbox,
                                            aria-activedescendant: option-2">
    <div id="real-listbox" role="listbox">
      <div id="option-1" role="option">Option 1</div>
      <div id="option-2" role="option">Option 2</div>
    </div>
  </template>
</fancy-listbox>
```
There's historically been a lot of talk about needing a holistic solution to clarify the path after a "good enough" solution, and I'm not sure what that would be in this case.

alice replied:
To clarify: this is asking whether the `role=combobox` is projected down on to the `<input>`?

Westbrook commented on reference-delegate-explainer.md:
````diff
@@ -0,0 +1,534 @@
+Declaring the mappings using the JavaScript API would look like:
+```js
+this.shadowRoot_.referenceDelegateMap['aria-controls'] = 'real-listbox';
+this.shadowRoot_.referenceDelegateMap['aria-activedescendant'] = 'option-2';
````
As `aria-activedescendant` would need to point to different IDs overtime, should we be explicit here, in preparation for spec text about this being a live reference in some way?

```js
this.shadowRoot_.referenceDelegateMap['aria-activedescendant'] = 'option-2';
// ...then later...
this.shadowRoot_.referenceDelegateMap['aria-activedescendant'] = 'option-1';
// ...etc.
```

behowell replied:
Yes, the intent is for these to be live references. I'll write that in explicitly.

nolanlawson replied:
Bikeshedding a bit, but it feels more natural to me in JS to use the camelCase format (ala the `ARIMixin`):

```js
this.shadowRoot_.referenceDelegateMap.ariaActiveDescendant = 'option-2';
```

(You might think `ReferenceDelegateMap` could just extend `ARIAMixin`, but that would require first solving https://github.com/w3c/aria/issues/1732, since `ariaActiveDescendant` and `ariaLabelledBy` and friends don't currently exist. 🙂 So maybe it's best to just define the 8 IDref camelCase formats just for `ReferenceDelegateMap`.)

nolanlawson replied:
`for` is also a funky case when it comes to the casing... Presumably this should be [`htmlFor`](https://developer.mozilla.org/en-US/docs/Web/API/HTMLLabelElement/htmlFor), i.e.  `referenceDelegateMap.htmlFor`?

Westbrook commented on reference-delegate-explainer.md:
```diff
@@ -0,0 +1,534 @@
+
+**Goals**
+
+- Create a mechanism for ID reference attributes like `aria-activedescendant` and `for` to refer to an element inside a component's shadow DOM.
```
Is there a path by which the `referenceDelegate` could receive other of the attributes from the host that are required to achieve accessible patterns (e.g. `role` or `aria-label`, etc.) without them being ID references?

Westbrook replied:
@alice re: https://github.com/WICG/aom/pull/207#discussion_r1425951582, yes. I ask it more clearly here, as I found the full combobox example at the end and deleted the other comment, sorry!

I realize that the explainer says that it _doesn't_ do this, but it seems like a thing that people would want it to do.

behowell replied:
I specifically left that out of scope of this proposal. This proposal only solves the problem of IDREF attributes that _target_ the host element, as that is the only "missing piece" blocking use of ARIA with the shadow DOM. It does not address attributes that are _on_ the host element itself.

It's possible to implement something similar to what you mention, today using custom observed attributes, and passing them to elements inside the shadow tree. Notably, the ARIAMixin attributes like `ariaActiveDescendantElement`, combined with custom attributes, allow you to refer OUT of the shadow DOM. That said, the current state isn't perfect: It doesn't let you use the "real" attributes like `aria-label` or `role`. Instead, you need to invent custom attributes to forward to the elements inside. But it is at least doable today.

In an earlier draft of this proposal, I had an appendix with a possible declarative syntax to achieve the same result. I removed it from this proposal in order to avoid distracting the main proposal. I've moved it to a separate file here: [Declarative syntax for referring OUT of the Shadow DOM](https://github.com/behowell/aom/blob/attribute-binding/attribute-binding-explainer.md). However, that is still very early and still needs a lot of work.

Westbrook commented on reference-delegate-explainer.md:
````diff
@@ -0,0 +1,534 @@
+
+```html
+<template id="t-combobox-input">
+  <input id="real-input" role="combobox" type="text" />
````
This requires that you have one input for your comboboxes and a different input for other things. Is there no path therein this could be a `<fancy-input>`?

behowell replied:
I think I addressed this in my reply here https://github.com/WICG/aom/pull/207#discussion_r1425977603, but let me know if you were asking about something else.

behowell replied:
I could update the combobox example to show how you could set the `role` of the input using a custom observed property on `fancy-input` (rather than a "new" `combobox-input` control). It's not new or specific to this proposal, but perhaps it helps show a more complete picture of referenceDelegate fitting in with other features.

behowell replied:
I've updated the combobox example to use a `<fancy-input>` that's not combobox-specific, instead of a specialized `<combobox-input>`. It's not fully-featured since it's just demoware, but hopefully it demonstrates the direction you could go to use a reusable custom input control for a combobox.

Westbrook commented on reference-delegate-explainer.md:
````diff
@@ -0,0 +1,534 @@
+```html
+<template id="t-fancy-combobox">
+  <combobox-input id="combo-input" 
+                  listbox="combo-listbox"> <!-- (1) -->
````
I wish there was a quality path towards normative API here, so that `<fancy-input>` could be consumed as a drop-in replacement of `<input>` in something like [the ARIA APG demos for combobox](https://www.w3.org/WAI/ARIA/apg/patterns/combobox/examples/combobox-autocomplete-both/#rps_label_textbox).

behowell commented on reference-delegate-explainer.md:
````diff
@@ -0,0 +1,534 @@
+
+The equivalent declarative attribute is `shadowrootreferencedelegatemap`, which is a comma-separated list of attribute to ID mappings.
+
+```html
+<input role="combobox"
+       aria-controls="fancy-listbox"
+       aria-activedescendant="fancy-listbox" />
+<fancy-listbox id="fancy-listbox">
+  <template shadowrootmode="closed"
+            shadowrootreferencedelegatemap="aria-controls: real-listbox,
+                                            aria-activedescendant: option-2">
+    <div id="real-listbox" role="listbox">
+      <div id="option-1" role="option">Option 1</div>
+      <div id="option-2" role="option">Option 2</div>
+    </div>
+  </template>
+</fancy-listbox>
+```
````
It looks like the comment you replied to has been deleted. But I answered a similar comment here: https://github.com/WICG/aom/pull/207#discussion_r1425977603

annevk commented on reference-delegate-explainer.md:
```diff
@@ -0,0 +1,534 @@
+
+#### Cons
+
+* The Reference Delegate feature is intended to support all attributes that use ID references. Thus, the only time a new attribute will be supported by Reference Delegate is when it is a completely new attribute in the HTML spec. There is no backwards compatibility concern, since no websites will be using the new attribute before is is supported.
```
I think this argument works when scoped to targeting, although presumably it should not be limited to ID references. What about `<label>`-sans-`for=""`?

However, before there was a suggestion that all kinds of other behavioral things would be delegated as well. E.g., focus would magically transfer. I guess that's off the table now?

behowell replied:
There's a [Nesting inside `<label>`](https://github.com/behowell/aom/blob/reference-delegate/reference-delegate-explainer.md#nesting-inside-label) section that currently says it'll work when nested in a label.

However, as I mentioned in a note in that section, that is the only part of this feature that doesn't directly work with IDREF attributes. Perhaps it's better to keep `referenceDelegate` scoped to deal _only_ with ID references?

Delegating other things isn't off the table; but it would be out of scope of this particular feature. Previous proposals have stalled in part due to scope creep, and it might be best to keep this feature focused on the one piece currently blocking ARIA from working across shadow roots.

alice replied:
I'm reluctantly in favour of leaving `<label>` wrapped out of scope - as much as I would love it to work, I agree that it feels like scope creep. 

It does seem like if it's included, it should be an explicit inclusion, to avoid opening a can of worms in figuring out how all the other types of nesting element behaviours would work and/or what would be in/out of scope.

annevk replied:
I think `<label>` without `for` should work. I don't know if any other elements that work like it, but I guess we should double check that.

behowell replied:
I've updated the proposal to say that label wrapping doesn't work without the `for` attribute. This keeps the feature focused only on ID references. I'm open to changing it back if there is strong feedback in favor of it.

nolanlawson commented on reference-delegate-explainer.md:
```diff
@@ -0,0 +1,534 @@
+    constructor() {
+      super();
+      this.shadowRoot_ = this.attachShadow({ mode: "closed" });
+      this.shadowRoot_.appendChild(
+        document.getElementById("t-fancy-input").content.cloneNode(true));
```
```suggestion
      this.shadowRoot_.innerHTML = '<input id="real-input">'
```

This does the same thing, but without needing the `<template>` above or to clone it.

I think it makes the example a bit clearer, since otherwise someone may be confused by the `real-input` ID elsewhere in the DOM, or the fact that this component is reaching into the global scope.

nolanlawson commented on reference-delegate-explainer.md:
```diff
@@ -0,0 +1,534 @@
+      if (attr === "activeitem") {
+        this.shadowRoot_.referenceDelegateMap['aria-activedescendant'] = value; // (2)
+      }
+    }
```
Minor quibble: I think a listbox component would provide better encapsulation by exposing methods to move the focus up/down rather than an `attributeChangedCallback` explicitly setting the index of the active item. The issue here is that the consumer of the listbox has to "know" how many items are in the listbox and which one is the active one (thus exposing internal details).

Whereas using methods (or events or what have you), the fancy input could just communicate to the listbox "Hey, the <kbd>↑</kbd> key was pressed," so the listbox could change its `referenceDelegateMap` internally without the consumer knowing how many items there are or which one is the active one.

behowell replied:
That does make sense. However, my goal was more to show how the referenceTargetMap works, rather than design a real listbox/combox. I'm inclined to leave it as-is to keep things simpler for the example.

nolanlawson commented on reference-delegate-explainer.md:
````diff
@@ -0,0 +1,534 @@
+    </div>
+  </template>
+</description-with-tooltip>
+```
````
If I understand this proposal correctly, it does still suffer from the bottleneck effect, right? For example, in this case, the consumer can't decide whether to refer to the `message` _or_ the `tooltip` – it can only refer to both at the same time.

Maybe this is an acceptable tradeoff (and as you say elsewhere, similar proposals have been delayed due to scope creep), but maybe it's worth calling out as an explicit non-goal.

behowell replied:
Yes, it does suffer from the [bottleneck effect](https://alice.pages.igalia.com/blog/how-shadow-dom-and-accessibility-are-in-conflict/#limitations-of-these-apis). However, I haven't been able to come up with a compelling example where the bottleneck effect is a limitation in real-world use.

The example in your comment isn't actually the bottleneck effect: the `description-with-tooltip` component _could_ provide some API to let the consumer pick whether to include the `message` or `tooltip` or both. The component would update its `referenceDelegateMap` based on that API.

The bottleneck effect would be if there were two separate elements outside of the `description-with-tooltip`, and one wanted its `aria-describedby` to be the `message`, and the other wanted its `aria-describedby` to be the `tooltip`. I can't think of a good reason why you'd do that...

behowell committed:
Rework introduction and background sections

alice commented on reference-delegate-explainer.md:
```diff
@@ -0,0 +1,562 @@
+* Requires new attributes in two places in order to work: E.g. `shadowrootreflectscontrols` on the shadow root _and_ `reflectariacontrols` on the target element.
+* When multiple elements are used for the same attribute, the author cannot control the order (the order is always the DOM order).
+
+### Use exported IDs instead of per-attribute targeting
```
I'd like to make one last pitch for this, and then I promise I'll let it go :) 

**The default `referenceDelegate` attribute (without the per-attribute targeting) supports a fairly well-understood and easy to articulate use case: the "semantic delegate" case.** This is a pattern [already widely used](https://github.com/alice/aom/blob/gh-pages/semantic-delegate.md#some-examples-of-custom-elements-from-component-libraries-which-use-this-pattern) in custom element libraries.

`referenceDelegate` supports the "semantic delegate" use case by solving the problem that, conceptually, the custom element _should_ be the target for certain IDREF attributes, but those attributes need to target the "delegate" element in practice. This preserves the encapsulation of the custom element by not requiring authors using the element to know anything about those implementation details. They can simply use the custom element as though it _is_ the delegate element[1].

**I think `referenceDelegateMap` is trying to solve _other problems_**, and I don't think it's solving them anywhere near as well.

The only obvious[2] case I can come up with for `referenceDelegateMap` is the `aria-activedescendant` case, like where `<input>` uses `aria-activedescendant` to refer to one of the `option`s inside the shadow root of a custom element representing a `listbox` in the [Combining `referenceDelegate` and `referenceDelegateMap`](https://github.com/behowell/aom/blob/reference-delegate/reference-delegate-explainer.md#combining-referencedelegate-and-referencedelegatemap) example. 

In this case, you _must know_ that there are `option` elements within the `<fancy-listbox>` which can be a valid target for `aria-activedescendant`, and you also _must know_ that the attribute will be forwarded correctly. This is a completely different case from the semantic delegate case, in which you _don't need to know anything about the internals_, only that you can treat the host as though it is what it "appears" to be.

**To me, `exportid` is a better fit for this type of problem**: where there is some element inside the shadow root that needs to be targeted explicitly, but for whatever reason needs to remain inside the shadow root. `exportid` allows you to specify an "API" for your element to allow references to elements inside its shadow root that you guarantee will exist, but need not reveal any more information about.

**`exportid` also gives us ample flexibility for unforeseen cases.** What I would especially like to avoid is a situation where we figure out all of the inherent complexity of supporting `referenceDelegateMap` (e.g. how to design its JavaScript API) and then realise that there actually are cases where the "bottleneck problem" is an issue, so we need to do something like `exportid` anyway. Not only would we then have spent extra effort in designing `referenceDelegateMap`, but now authors have to choose which one to use for their largely overlapping use cases.

---

[1] Leaving aside that you can't use IDREF attributes on the host and also have them forwarded, which was deliberately left out of scope here to keep the scope as minimal as possible.

[2] I would really like to understand what other cases there are! @nolanlawson you mentioned doing an audit of your component library recently; perhaps you have some examples? The example under [Delegating to multiple elements](https://github.com/behowell/aom/blob/reference-delegate/reference-delegate-explainer.md#delegating-to-multiple-elements) is interesting, but seems like it would equally be a problem in a non-shadow root world. That is, assuming the content inside the shadow root in this example is intended to be "unknown" (i.e. user-generated or similar) to the page author, the author would still have no way in practice to correctly set `aria-describedby` on the `<input>` even if that content were in light DOM. And if it was user-generated, how would you correctly set the `referenceDelegateMap`? But, maybe this is all wishful thinking on my part :)

nolanlawson replied:
@alice The `ariaActiveDescendant` case is definitely the strongest one I've seen. The fact that you want an `ariaControls` as well as an `ariaActiveDescendant` pointing into the same shadow root came up twice in my component audit (both were comboboxes).

Aside from that, every use case I've seen for any kind of cross-root references could be solved by either a "semantic delegate" type solution, or [element reflection](https://chromestatus.com/feature/6244885579431936) rather than reference delegates – including the tricky case of multiple elements inside one shadow root sharing the same ARIA attribute (see our [lightning-input](https://developer.salesforce.com/docs/component-library/bundle/lightning-input/specification) and its `date-aria-*` and `time-aria-*` attrs, which allow for a separate date element and time element to each point to separate elements in their shadow-including-ancestor using `controls`/`describedBy`/`labelledBy`). This has led me to believe that the "bottleneck effect" is not such a dealbreaker for the current proposal.

I definitely agree though that it feels "weird" to have (say) the listbox be both the target for `controls` as well as `activeDescendant`. But I would still be happy with this proposal since it gives us a path forward – perhaps userland solutions can make it more ergonomically appealing. 🙂

behowell committed:
Add description of live references, and change referenceDelegateMap to camelCase

behowell committed:
Update combobox example to use generic/reusable fancy-input

asyncliz commented on reference-delegate-explainer.md:
```diff
@@ -0,0 +1,555 @@
+- This feature does not solve the [bottleneck effect](https://alice.pages.igalia.com/blog/how-shadow-dom-and-accessibility-are-in-conflict/#limitations-of-these-apis). It is difficult to find a compelling real-world example where this is a problem.
+- This does not affect the how attributes set on the host element work. For example, this does not tackle the problem of forwarding `role` or `aria-label`, etc. from the host element to an element inside.
+
+### The `referenceDelegate` attribute
```
Do you think naming these attributes `delegatesReference` and `delegatesReferenceMap` makes sense alongside `delegatesFocus`?

behowell replied:
I like the name `delegatesReference`, and that it is parallel to `delegatesFocus`. However, it does _sound_ more like it should be a boolean true/false, rather than the ID of the target. (E.g. it should be `delegatesReferenceTo` if it's the ID of the target, but I don't like that option). The name `delegatesReferenceMap` is also a little awkward IMO.

I took another look at the names used for [IDREF attributes](https://github.com/behowell/aom/blob/reference-delegate/reference-delegate-explainer.md#supported-attributes) to see if there was possible inspiration there... The naming of the ARIA attributes isn't very consistent: some are verbs (`aria-controls`), some are nouns (`aria-activedescendant`), and some are prepositional phrases (`aria-labelledby`). However, several of the newer attributes follow the pattern of ending with "target": `popovertarget`, `invoketarget`, `interesttarget`.

So, another possible option is `referenceTarget` / `referenceTargetMap`. WDYT?

asyncliz replied:
I like that since it helps enforce a common pattern using the "target" suffix.

Another option could be `idReferenceTarget`/`idReferenceTargetMap` or `idRefTarget`/`idRefTargetMap` to emphasize the relationship to IDREF attributes.

behowell replied:
I've updated the proposal to the `referenceTarget`/`referenceTargetMap` name.

behowell committed:
Rename to Reference Target

behowell committed:
Spellcheck and edits for clarity

behowell committed:
Split into Phase 1/Phase 2, and rearrange some sections for clarity

behowell committed:
Edits to address PR feedback and for clarity

behowell committed:
Run formatter

behowell committed:
Rename file to `reference-target-explainer.md`

behowell committed:
Edits for clarity to the Introduction and Background

behowell commented:
Apologies this has been sitting around for a while. I've made some updates to this proposal. Here's a summary:
1. Renamed the feature to "Reference Target" instead of "Reference Delegate". See https://github.com/WICG/aom/pull/207#discussion_r1445354205
2. Broke proposal into two phases. Phase 1 includes only the `referenceDelegate` attribute to try to get that feature in faster. Phase 2 can come later to give more bake time for the `referenceDelegateMap` attribute, or some alternative.
3. Made updates based on PR feedback, and for clarity.

It would be great to get this explainer merged into the AOM repo once people have had a chance to look at the edits. Thank you for the reviews!

cc @alice, @nolanlawson, @Westbrook, @annevk, @asyncLiz

nolanlawson commented on reference-target-explainer.md:
```diff
@@ -0,0 +1,611 @@
+
+- This is scoped to only solve the problem of referring _into_ the shadow DOM. It relies on ARIAMixin to refer _out_ of the shadow DOM.
+- This feature does not solve the [bottleneck effect](https://alice.pages.igalia.com/blog/how-shadow-dom-and-accessibility-are-in-conflict/#limitations-of-these-apis). It is difficult to find a compelling real-world example where this is a problem.
+- This does not affect the how attributes set on the host element work. For example, this does not tackle the problem of forwarding `role` or `aria-label`, etc. from the host element to an element inside.
```
```suggestion
- This does not affect how attributes set on the host element work. For example, this does not tackle the problem of forwarding `role` or `aria-label`, etc. from the host element to an element inside.
```

nolanlawson commented on reference-target-explainer.md:
```diff
@@ -0,0 +1,611 @@
+      constructor() {
+        super();
+        this.shadowRoot_ = this.attachShadow({ mode: "closed" });
+        this.shadowRoot_.innerHTML = `<input id="real-input" />`;
```
```suggestion
        this.shadowRoot_.innerHTML = `<input id="real-input">`;
```

`<input>s` are self-closing in HTML.

nolanlawson commented on reference-target-explainer.md:
```diff
@@ -0,0 +1,611 @@
+    shadowrootmode="closed"
+    shadowrootreferencetarget="real-input"
+  >
+    <input id="real-input" />
```
```suggestion
    <input id="real-input">
```

See above.

nolanlawson commented on reference-target-explainer.md:
````diff
@@ -0,0 +1,611 @@
+    </div>
+  </template>
+</fancy-listbox>
+```
````
This is very powerful. This would support both the usecase where 1) the component author wants to explicitly limit the `aria-*` attributes that can be used to target specific elements, as well as 2) the case where the component author doesn't care and wants to just have any wildcard `aria-*` attribute point at a specific element. It also does so in such a way that doesn't force usecase #<span></span>2 down the explicit path if they need to set aside one `aria-*` attribute for a particular target (as in your `aria-activedescendant` example). I like it!

nolanlawson commented on reference-target-explainer.md:
```diff
@@ -0,0 +1,611 @@
+- `HTMLLabelElement.control`
+- _(This list is not exhaustive)_
+
+These will _always_ refer to the **host** element that they're targeting, and _never_ the referenceTarget element directly. This behavior maintains the design that an [IDL attribute with type Element](https://html.spec.whatwg.org/multipage/common-dom-interfaces.html#reflecting-content-attributes-in-idl-attributes:element) can only refer to an element that is a descendant of a [shadow-including an ...
```
I like this. It exposes the minimal API surface while also (potentially) allowing authors with full access to open shadow roots to traverse the DOM and build the equivalent of `composedPath`/`activeElement`-drilling/etc.

nolanlawson commented on reference-target-explainer.md:
```diff
@@ -0,0 +1,611 @@
+    constructor() {
+      super();
+      this.shadowRoot_ = this.attachShadow({ mode: "closed" });
+      this.shadowRoot_.innerHTML = `<input id="real-input" />`;
```
```suggestion
      this.shadowRoot_.innerHTML = `<input id="real-input">`;
```

See above.

nolanlawson reviewed (approved):
Looks great to me. I especially like how it's broken into two phases, with the first phase covering the most common use case.

behowell committed:
Apply suggestions from code review

Co-authored-by: Nolan Lawson <nolan@nolanlawson.com>

Westbrook reviewed (approved):
Very excited to see this continue to move forward!!

behowell committed:
Add warning to exportid explainer.

behowell commented on exportid-explainer.md:
```diff
@@ -1,3 +1,6 @@
+> [!CAUTION]  
+> The Export ID proposal is no longer planned. It has been replaced by the [Reference Target](reference-target-explainer.md) proposal.
+
```
Adding this warning to the exportid explainer instead of deleting it, since (a) it's been linked from other places and can serve as a redirect to the Reference Target explainer, and (b) some ideas from it may still be useful to reference as part of the discussion on referenceTargetMap.

behowell replied:
But let me know if there's another preferred way to do this for the AOM repo!

alice replied:
I think this is good :)

asyncliz reviewed (approved):
This looks great!! Thank you so much for all the work in pushing this forward 🥳

alice commented on reference-target-explainer.md:
```diff
@@ -0,0 +1,611 @@
+However, there are still missing pieces that prevent a web component from truly being a drop-in replacement for a built-in, including:
+
+1. Can't create ID reference links to elements inside the shadow DOM.
+2. Can't use built-in attributes like `aria-label` or `role` on the host and have them apply to an element inside the shadow DOM.
```
```suggestion
2. Can't use built-in attributes like `aria-label` or `role` on the host and have them apply to an element inside the shadow root.
```

alice commented on reference-target-explainer.md:
```diff
@@ -0,0 +1,611 @@
+
+1. Can't create ID reference links to elements inside the shadow DOM.
+2. Can't use built-in attributes like `aria-label` or `role` on the host and have them apply to an element inside the shadow DOM.
+3. Non-trivial amount of code required to hook up custom attributes on the host to ARIAMixin attributes on an element inside the shadow DOM.
```
```suggestion
3. Non-trivial amount of code required to hook up custom attributes on the host to ARIAMixin attributes on an element inside the shadow root.
```

alice commented on reference-target-explainer.md:
```diff
@@ -0,0 +1,611 @@
+
+However, there are still missing pieces that prevent a web component from truly being a drop-in replacement for a built-in, including:
+
+1. Can't create ID reference links to elements inside the shadow DOM.
```
```suggestion
1. Can't create ID reference links to elements inside a [shadow tree](https://dom.spec.whatwg.org/#shadow-trees).
```

alice commented on reference-target-explainer.md:
```diff
@@ -0,0 +1,611 @@
+- [Phase 1](#phase-1) adds the ability to designate a single element as the target for _all_ IDREF properties that refer to the host.
+- [Phase 2](#phase-2) adds a way to re-target specific properties (like `aria-activedescendant`) to refer a separate element.
+
+The goal of breaking it into phases is to get the simpler and less controversial ideas working first. The solutions to Phase 2 are more complex and may need more discussion before they are ready.
```
```suggestion
The goal of breaking it into phases is to get the simpler syntax and simpler use cases working first. The solutions to Phase 2 are more complex and may need more discussion before they are ready.
```

"Controversial" depends on who you ask :)

alice commented on reference-target-explainer.md:
```diff
@@ -0,0 +1,611 @@
+
+#### Cons
+
+- Exposes some of the internal details of a control and does not give a way for the control to encapsulate those details.
```
I'm still not really clear on how this is the case any more than it is for `::part()`.

behowell replied:
I added a sub-bullet to this con to capture the feedback:
>   * This may not be a dealbreaker: the [`::part()`](https://drafts.csswg.org/css-shadow-parts/#part) CSS selector also has a similar drawback for CSS styles, but it still is a standard and a useful feature for styles.

behowell committed:
Apply suggestions from code review

Co-authored-by: Alice <95208+alice@users.noreply.github.com>

behowell committed:
Update reference-target-explainer.md

Add `anchor` to the list of supported attributes, and add a note about `::part()` to the "Cons" list of ExportID.

behowell commented:
Thank you for the feedback everyone! I'm going to merge this PR to get the explainer checked in. Further feedback can be in the form of an issue in the aom repo. Thanks!

alice commented:
Nice one! Thanks for all your work on this!

EisenbergEffect commented:
Just saw that this was merged. Awesome work!
